lib/string.c

```diff
@@ -119,39 +114,63 @@ int64_t meth_Str_unicode_length(const struct class_Str *s)
 	return res;
 }
 
-// TODO: for most uses, it is inefficient to allocate a new object
-struct class_Str *meth_Str_slice(const struct class_Str *s, int64_t start, int64_t end)
+static struct class_Str *slice_from_start(struct class_Str *s, size_t len)
 {
-	int64_t len = strlen(s->str);
-	if (start < 0)
-		start = 0;
-	if (start > len)
-		start = len;
-	if (end < 0)
-		end = 0;
-	if (end > len)
-		end = len;
-
-	if (start >= end)
-		return cstr_to_string("");
-	if (is_utf8_continuation_byte(s->str[start]))
-		panic_printf("can't slice string in the middle of continuation byte");
-
-	if (end == len)
-		return cstr_to_string(&s->str[start]);
-	if (is_utf8_continuation_byte(s->str[end]))
-		panic_printf("can't slice string in the middle of continuation byte");
-
-	struct class_Str *res = alloc_string(end - start);
-	memcpy(res->str, &s->str[start], end - start);
-	res->str[end - start] = '\0';
+	assert(strlen(s->str) >= len);
+	// TODO: avoid temporary allocation
+	char *tmp = malloc(len + 1);
+	assert(tmp);
+	memcpy(tmp, s->str, len);
+	tmp[len] = '\0';
+	struct class_Str *res = cstr_to_string(tmp);
+	free(tmp);
 	return res;
 }
 
-int64_t string_find_internal(const struct class_Str *s, const struct class_Str *sub)
+// TODO: avoid allocations
+struct class_Str *string_get_first_char(struct class_Str *s)
+{
+	assert(s->str[0] != '\0');
+	size_t n = 1;
+	while (is_utf8_continuation_byte(s->str[n]))
+		n++;
+	return slice_from_start(s, n);
+}
+
+// TODO: for most uses, it is inefficient to allocate a new object
+struct class_Str *string_remove_prefix(struct class_Str *s, struct class_Str *pre)
+{
+	size_t n = strlen(pre->str);
+	if (strlen(s->str) >= n && memcmp(s->str, pre->str, n) == 0)
+		return cstr_to_string(s->str + strlen(pre->str));
+	incref(s);
+	return s;
+}
+
+struct class_Str *string_remove_suffix(struct class_Str *s, struct class_Str *suf)
+{
+	size_t slen=strlen(s->str), suflen=strlen(suf->str);
+	if (slen >= suflen && memcmp(s->str + slen - suflen, suf->str, suflen) == 0)
```

Comment:
perhaps a bit less repetitive:
```suggestion
	if (slen >= suflen && strcmp(s->str + slen - suflen, suf->str) == 0)
```